lib/3scale/backend/storage_async/client.rb

```diff
@@ -51,13 +54,39 @@ def pipelined(&block)
           # This replaces the client with a Pipeline that accumulates the Redis
           # commands run in a block and sends all of them in a single request.
 
-          pipeline = Pipeline.new
-          block.call pipeline
-          pipeline.run(@redis_async)
+          ensure_connected do |conn|
+            pipeline = Pipeline.new
+            block.call pipeline
+            pipeline.run(conn)
+          end
+
         end
 
         def close
-          @redis_async.close
+          @redis_async&.close
+          @redis_async = nil
+        end
+
+        def ensure_connected
+          attempt = 0
+          begin
+            @redis_async ||= AsyncRedis::Client.connect(@opts)
+
+            if block_given?
+              yield @redis_async
+            else
+              @redis_async
+            end
+          rescue *CONNECTION_ERRORS => e
+            close
+
+            if attempt < @opts[:reconnect_attempts]
+              attempt += 1
+              retry
```

Comment:
Yeah, I'm sorry, sometimes when I say requests I mean the HTTP requests received by the listener, and some other times I mean the requests we send to Redis... my fault.

As you say, this retry will work if the connection is reestablished when the second attempt is made, but since the retry happens immediately, most of times it will fail again, and raise the exeception. I'm implementing this because it's how `redis-client` does it in sync mode, and I think it's good to accept the same parameters and behave as equal as possible in both sync and async modes. However, this retry won't change anything most of times IMO.

Reply:
So HTTP requests will fail while connection is down, but then queue will be retried? Or still not fully get it?

Reply:
Well, first of all, this client is common for listener and worker. In both cases, when the connection to redis fails is retried `:reconnect_attempts` times.

Then, after retrying `:reconnect_attempts` times, the exception is raised, and it's caught by the caller.

- When the caller is the listener, it returns 403 to the client.
- When the is the worker, it waits 5 seconds and tries again, and repeats this forever.

After this PR, this behavior is the same in sync and async modes.